8Knot/_login.py

```diff
@@ -0,0 +1,216 @@
+            password=os.getenv("REDIS_PASSWORD", ""),
+        )
+        try:
+            users_cache.ping()
+        except redis.exceptions.ConnectionError:
+            logging.error("LOGOUT: Could not connect to users-cache.")
+            return redirect("/")
+
+        if current_user.is_authenticated:
+            c_id = current_user.get_id()
+            users_cache.delete(c_id)
+            logout_user()
+            logging.warning(f"USER {c_id} LOGGED OUT")
+        else:
+            logging.warning("TRIED TO LOG OUT")
+        return redirect("/")
+
+    @server.route("/login/")
+    def oauth2_authorize():
+        users_cache = redis.StrictRedis(
+            host="redis-users",
+            port=6379,
+            password=os.getenv("REDIS_PASSWORD", ""),
+        )
+        try:
+            users_cache.ping()
+        except redis.exceptions.ConnectionError:
+            logging.error("LOGIN: Could not connect to users-cache.")
+            return redirect("/")
+
+        provider = os.environ.get("OAUTH_CLIENT_NAME")
+
+        if not current_user.is_anonymous:
+            return redirect("/")
+
+        provider_data = current_app.config["OAUTH2_PROVIDERS"].get(provider)
+        if provider_data is None:
+            abort(404)
+
+        # generate a random string for the state parameter
+        session["oauth2_state"] = secrets.token_urlsafe(16)
+
+        # create a query string with all the OAuth2 parameters
+        qs = urlencode(
+            {
+                "client_id": provider_data["client_id"],
+                # "redirect_uri": url_for("oauth2_callback", _external=True),
+                "response_type": "code",
+                # "state": session["oauth2_state"],
+            }
+        )
+
+        # redirect the user to the OAuth2 provider authorization URL
+        return redirect(provider_data["authorize_url"] + "?" + qs)
+
+    @server.route("/authorize/")
+    def oauth2_callback():
+        users_cache = redis.StrictRedis(
+            host="redis-users",
+            port=6379,
+            password=os.getenv("REDIS_PASSWORD", ""),
+        )
+        try:
+            users_cache.ping()
+        except redis.exceptions.ConnectionError:
+            logging.error("AUTHORIZE: Could not connect to users-cache.")
+            return redirect("/")
+
+        provider = os.environ.get("OAUTH_CLIENT_NAME")
+
+        if not current_user.is_anonymous:
+            return redirect("/")
+
+        provider_data = current_app.config["OAUTH2_PROVIDERS"].get(provider)
+        if provider_data is None:
+            abort(404)
+
+        # if there was an authentication error, flash the error messages and exit
+        if "error" in request.args:
+            for k, v in request.args.items():
+                if k.startswith("error"):
+                    flash(f"{k}: {v}")
+            return redirect("/")
+
+        # make sure that the state parameter matches the one we created in the
```

Comment:
are these comments still relivent?

Reply:
Yeah, Augur could implement state mirroring in the future, so this part of the implementation may become relevant. This comment also saves time because it shows the choice of not monitoring state mirroring.